### PR TITLE
[DONT MERGE] Added Near Cache test for TransactionalMap

### PR DESCRIPTION
**Pull request by Donnerbart**
Found untested code in the `TransactionalMapProxySupport` regarding Near Cache usage.

### CODE REVIEW

File: hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        assertTrue("Near Cache size should be > 0 but was " + nearCacheSize, nearCacheSize > 0);
+
+        transactionContext = client.newTransactionContext();
+        transactionContext.beginTransaction();
+        map = transactionContext.getMap(mapName);
+        for (Integer key : map.keySet()) {
+            map.remove(key);
+        }
+        transactionContext.commitTransaction();
+        assertTrueEventually(
```
Comment (jerrinot) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: this would deserve its own method. `assertNearCacheEmpty()` <- it makes the test more readable.

Comment (Donnerbart) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: Done

**Comment by Donnerbart** [comment]
The duplication of Near Cache test code is an issue in general. The problem is that most of the member/client tests are configured quite differently, so this would be a bigger cleanup.

In this specific case I'm mostly worried that one test configuration always fails to clean the Near Cache via the `TransactionalMap`.

**Comment by Donnerbart** [comment]
This one test is really constantly failing, locally and on the build machine...
`testBasicUsage()` with `batchInvalidationEnabled:false`
```
java.lang.AssertionError: Near Cache size on instance _hzInstance_6_dev should be 0 but was 3307 
Expected :0
Actual   :3307
```
It's the Near Cache of the first instance, which removes the keys (since the `TransactionalMap` has no `clear()` method).

File: hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    protected final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
+
+    @After
+    public void tearDown() {
+        hazelcastFactory.shutdownAll();
+    }
+
+    @Test
+    public void test_whenEmptyMap_thenPopulatedNearCacheShouldReturnNull_neverNULL_OBJECT() {
```
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: the last part of this method name is very confusing. 

test_whenEmptyMap_thenPopulatedNearCacheShouldReturnNull_butNeverNULLOBJECT

?

Now it reads as a sentence. And you can get rid of test_

File: hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            // populate Near Cache
+            assertNull(map.get(i));
+            // fetch value from Near Cache
+            assertNull(map.get(i));
+        }
+        transactionContext.commitTransaction();
+    }
+
+    @Test
+    public void testBasicUsage() {
```
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: Personally I'm not crazy about such tests since you do, verify, do, verify, do.

Each of these steps deserves its own test.

File: hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        map = transactionContext.getMap(mapName);
+        for (Integer key : map.keySet()) {
+            map.remove(key);
+        }
+        transactionContext.commitTransaction();
+        assertNearCacheEmptyEventually(mapName, client);
+    }
+
+    @Override
+    protected Config getConfig() {
```
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: I would not call this method 'getConfig' since you are not 'getting' something, you are creating a new config. So newConfig or createConfig would be better describing the purpose of this method.

File: hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        for (int i = 0; i < mapSize; i++) {
+            map.put(i, i);
+        }
+    }
+
+    private ClientConfig newClientConfig() {
+        return new ClientConfig();
+    }
+
+    private HazelcastInstance getNearCachedClient(String mapName, Config config, NearCacheConfig nearCacheConfig) {
```
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTxTest.java]: Same as above. newNearCacheClient or createNearCacheClient.

File: hazelcast/src/test/java/com/hazelcast/map/nearcache/NearCacheTxTest.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    @Test
+    public void test_whenEmptyMap_thenPopulatedNearCacheShouldReturnNull_neverNULL_OBJECT() {
+        int size = 1;
+        String mapName = randomMapName();
+
+        Config config = getConfig();
+        config.getMapConfig(mapName).setNearCacheConfig(newNearCacheConfig());
+
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/map/nearcache/NearCacheTxTest.java]: You can shorten this to

HazelcastInstance instance = createHazelcastInstance(config)

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by hasancelik** [comment]
run-lab-run

**Comment by Donnerbart** [comment]
run-lab-run
